WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/EditShippingLabelPaymentViewModel.kt

```diff
@@ -67,53 +74,80 @@ class EditShippingLabelPaymentViewModel @Inject constructor(
         viewState = viewState.copy(paymentMethods = paymentMethodsModels)
     }
 
+    fun onAddPaymentMethodClicked() {
+        triggerEvent(AddPaymentMethod)
+    }
+
     fun onDoneButtonClicked() {
         launch {
-            viewState = viewState.copy(showSavingProgressDialog = true)
-            val selectedPaymentMethod = viewState.paymentMethods.find { it.isSelected }!!.paymentMethod
-            val result = shippingLabelRepository.updatePaymentSettings(
-                selectedPaymentMethodId = selectedPaymentMethod.id,
-                emailReceipts = viewState.emailReceipts
-            )
-            viewState = viewState.copy(showSavingProgressDialog = false)
-
-            if (result.isError) {
-                triggerEvent(ShowSnackbar(string.shipping_label_payments_saving_error))
-            } else {
-                triggerEvent(ExitWithResult(selectedPaymentMethod))
+            val selectedPaymentMethod =
+                viewState.paymentMethods.find { it.isSelected }!!.paymentMethod
+
+            val requiresSaving = selectedPaymentMethod.id != viewState.currentAccountSettings?.selectedPaymentId ||
+                    viewState.emailReceipts != viewState.emailReceipts
+
+            if (requiresSaving) {
+                viewState = viewState.copy(showSavingProgressDialog = true)
+                val result = shippingLabelRepository.updatePaymentSettings(
+                    selectedPaymentMethodId = selectedPaymentMethod.id,
+                    emailReceipts = viewState.emailReceipts
+                )
+                viewState = viewState.copy(showSavingProgressDialog = false)
+
+                if (result.isError) {
+                    triggerEvent(ShowSnackbar(R.string.shipping_label_payments_saving_error))
+                    return@launch
+                }
             }
+            triggerEvent(ExitWithResult(selectedPaymentMethod))
         }
     }
 
     fun onBackButtonClicked() {
         triggerEvent(Exit)
     }
 
+    fun refreshData() {
+        launch {
+            loadPaymentMethods(forceRefresh = false)
+        }
+    }
+
+    fun onPaymentMethodAdded() {
+        launch {
+            val countOfCurrentPaymentMethods = viewState.paymentMethods.size
+            loadPaymentMethods(forceRefresh = true)
+            if (viewState.uiState == Success &&
+                    viewState.paymentMethods.size == countOfCurrentPaymentMethods + 1) {
+                triggerEvent(ShowSnackbar(R.string.shipping_label_payment_method_added))
+            }
+        }
+    }
+
     @Parcelize
     data class ViewState(
-        val isLoading: Boolean = false,
-        private val currentAccountSettings: ShippingAccountSettings? = null,
+        val uiState: UiState? = null,
```

Comment:
🔧  It's quite confusing to me what's the difference between `View` and `UI` states. Can we please think of some other naming? I see that it's sate of the "load data" operations that are performed, maybe something like 
```suggestion
        val loadDataState: LoadDataState? = null,
```

Reply:
I'm not really convinced 🙂, but I can change it.

Reply:
done in d24cc6e